### ISSUE TITLE
Introduction article Cross sections of Posterior plot has legend without patches

### ISSUE THREAD

**Issue by nipunbatra** [pyro-ppl/pyro#3016]
The [introduction page](http://pyro.ai/examples/intro_long.html) in Cell 23 has contour plots showing the cross sections of Posterior.

![intro_long_71_0](https://user-images.githubusercontent.com/60985/154836970-a2bfdde9-9cee-4d4a-8dec-112409678a1c.png)

The legend does not have the appropriate patches to help distinguish between the two guides (SVI Diagonal Normal and SVI MV Normal)

I think there are a couple of ways to solve this. 

### Approach 1 (using Pandas and `hue=` in Seaborn)

```python
svi_samples_df = pd.DataFrame({k: v.detach().cpu().numpy() for k, v in samples.items()})
svi_mvn_samples_df = pd.DataFrame({k: v.detach().cpu().numpy() for k, v in mvn_samples.items()})

svi_samples_df['Guide'] = 'Diagonal Normal'
svi_mvn_samples_df['Guide'] = 'Multivariate Normal'

svi_all_df = pd.concat([svi_samples_df, svi_mvn_samples_df])
svi_all_df = svi_all_df.reset_index()

fig, axs = plt.subplots(nrows=1, ncols=2, figsize=(12, 6))
fig.suptitle("Cross-sections of the Posterior Distribution", fontsize=16)
sns.kdeplot(data = svi_all_df, x = "bA", y="bR", ax=axs[0], hue='Guide',  shade=True, alpha=0.5)
sns.kdeplot(data = svi_all_df, x = "bR", y="bAR", ax=axs[1], hue='Guide', shade=False)
```

This produces the following (I shaded the two subplots differently on purpose)

![1](https://user-images.githubusercontent.com/60985/154837047-807b2458-8490-46b7-9789-bee55d667c66.png)

### Approach 2: Create the patch for the legend

```python
svi_samples = {k: v.detach().cpu().numpy() for k, v in samples.items()}
svi_mvn_samples = {k: v.detach().cpu().numpy() for k, v in mvn_samples.items()}

fig, axs = plt.subplots(nrows=1, ncols=2, figsize=(12, 6))
fig.suptitle("Cross-sections of the Posterior Distribution", fontsize=16)
sns.kdeplot(svi_samples["bA"], y=svi_samples["bR"], ax=axs[0], )
sns.kdeplot(svi_mvn_samples["bA"], y=svi_mvn_samples["bR"], ax=axs[0], shade=True )
axs[0].set(xlabel="bA", ylabel="bR", xlim=(-2.5, -1.2), ylim=(-0.5, 0.1))

sns.kdeplot(svi_samples["bR"], y=svi_samples["bAR"], ax=axs[1], )
sns.kdeplot(svi_mvn_samples["bR"], y=svi_mvn_samples["bAR"], ax=axs[1], shade=True )
axs[1].set(xlabel="bR", ylabel="bAR", xlim=(-0.45, 0.05), ylim=(-0.15, 0.8))


for label, color in zip(["SVI (Diagonal Normal)", "SVI (Multivariate Normal)"], sns.color_palette()[:2]):
    plt.plot([], [],
                label=label, color=color)
fig.legend(loc='upper right')
```

![2](https://user-images.githubusercontent.com/60985/154837928-7b497b5f-a0cc-46af-a77c-e1ff71f17b85.png)

#### Bandwidth adjustment
Another thing to consider might be the bandwidth adjustment param in KDE plot. Here is the plot with `bw_adjustment = 4`.

![3](https://user-images.githubusercontent.com/60985/154838550-99f73cf7-5b2d-4f2d-b78b-310bc5cd8044.png)

```python
svi_samples = {k: v.detach().cpu().numpy() for k, v in samples.items()}
svi_mvn_samples = {k: v.detach().cpu().numpy() for k, v in mvn_samples.items()}

fig, axs = plt.subplots(nrows=1, ncols=2, figsize=(12, 6))
fig.suptitle("Cross-sections of the Posterior Distribution", fontsize=16)
sns.kdeplot(svi_samples["bA"], y=svi_samples["bR"], ax=axs[0], bw_adjust=4 )
sns.kdeplot(svi_mvn_samples["bA"], y=svi_mvn_samples["bR"], ax=axs[0], shade=True, bw_adjust=4 )
axs[0].set(xlabel="bA", ylabel="bR", xlim=(-2.5, -1.2), ylim=(-0.5, 0.1))

sns.kdeplot(svi_samples["bR"], y=svi_samples["bAR"], ax=axs[1],bw_adjust=4 )
sns.kdeplot(svi_mvn_samples["bR"], y=svi_mvn_samples["bAR"], ax=axs[1], shade=True, bw_adjust=4 )
axs[1].set(xlabel="bR", ylabel="bAR", xlim=(-0.45, 0.05), ylim=(-0.15, 0.8))


for label, color in zip(["SVI (Diagonal Normal)", "SVI (Multivariate Normal)"], sns.color_palette()[:2]):
    plt.plot([], [],
                label=label, color=color)
fig.legend(loc='upper right')
```

The benefit of `bw_adjustment` might be to make the contours smoothers. It would be easier to identify that the Diagonal Normal is indeed axis-aligned while the MVN has non-zero covariance terms and thus not axis aligned. 

Let me know if you'd like me to make a PR and if so with which approach (1 or 2) and with or without bandwidth adjustment. 

I'm assuming I'd have to modify [this notebook](https://github.com/pyro-ppl/pyro/blob/dev/tutorial/source/intro_long.ipynb). Would the PR need anything else?

**Comment by fritzo** [comment]
Hi @nipunbatra, thanks for pointing this out. I think approach 2 would be good. Bandwidth adjustment looks great. Can we keep the entire ellipses within the box?
